### PR TITLE
Add Infection

### PR DESCRIPTION
**Pull request by antonio-gg-dev**
## 📚 Description

Out of curiosity, I have installed and configured Infection in the project. It is a library that allows applying the mutation testing methodology in PHP. This methodology makes small modifications to the production code to try to detect whether the test suite detects these changes or not. Each of these small changes should always be detected by one of our tests. If not, the library reports it so that we can act accordingly.

This library thus makes our tests better, helping us achieve 100% coverage and ensuring that we test all edge cases. However, like everything else, being an automated process, it can also produce false positives. Therefore, it is up to our discretion to decide whether to fix or ignore these failures.

I launched the command to take a look and the truth is that we got a pretty good results. However, it still detected that one mutation had no coverage, 14 were not detected by any test, and 8 timed out. Nonetheless, we have some enviable metrics!

imagen
imagen

I'm creating this draft to work on it, so you can take a look and decide whether to incorporate this methodology (perhaps to apply it to the other Gacela modules) and to check the errors it has detected. We can also add it as a step in the CI pipeline.

## 🔖 Changes

- I have installed Infection as a development dependency.
- I have added Infection as new script in the composer.json file, in such a way that it launches as many threads as cores has the computer where the command is executed.
- I have added this new script as the last step in the composer ctal script.
- I have added a basic configuration for Infection to work for now, and we will have to fine-tune it gradually.

## 🤔 Considerations

Since my computer has 16 cores, the script's execution time is exactly 10 seconds, which is the default timeout amount. Currently, we have 8 timeouts, so on computers with less than 8 cores, the execution time will be between 20 and 80 seconds.

We need to study whether these timeouts are something we can fix (or ignore), so that we ensure they don't potentially add 10 seconds each on slower computers (those with a single core). However, since our test suite is so fast, another option we have is to drastically reduce the timeout amount. Nevertheless, as I mentioned earlier, we need to consider what to do in this situation.

### CODE REVIEW

**Comment by Chemaclass** [comment]
I love the idea of incorporating mutation testing in this project. I want tests to test my tests! 💡🥳

File: src/Router/Entities/RouteParams.php [hunk 1]
```diff
@@ -49,20 +48,16 @@ private function getParams(): array
             ->getParameters();
 
         foreach ($actionParams as $actionParam) {
-            /** @var string|null $paramType */
-            $paramType = null;
-
-            if ($actionParam->getType() && is_a($actionParam->getType(), ReflectionNamedType::class)) {
-                $paramType = $actionParam->getType()->getName();
-            }
+            /**@var string|null $paramType */
```
Comment (Chemaclass) [inline src/Router/Entities/RouteParams.php]:
```suggestion
            /** @var string|null $paramType */
```

File: src/Router/Entities/RouteParams.php [hunk 2]
```diff
@@ -50,20 +50,17 @@ private function getParams(): array
...
-                'int' => (int)($pathParams[$paramName] ?? 0),
-                'float' => (float)($pathParams[$paramName] ?? 0.0),
-                'bool' => (bool)json_decode($pathParams[$paramName] ?? '0'),
-                null => null,
+                'string' => $pathParams[$paramName],
+                'int' => (int)$pathParams[$paramName],
+                'float' => (float)$pathParams[$paramName],
+                'bool' => (bool)json_decode($pathParams[$paramName]),
+                null => throw UnsupportedParamTypeException::nonTyped(),
+                default => throw UnsupportedParamTypeException::fromType($paramType),
```
Comment (antonio-gg-dev) [inline src/Router/Entities/RouteParams.php:60]: @Chemaclass @JesusValera I need help there, as result of changes on commit https://github.com/gacela-project/router/pull/29/commits/1bbaa4872ae0d988d3038b7f9b63e8f9daa16470 phpstan is throwing this error `Ignored error pattern #Match expression does not handle remaining value: string# was not matched in reported errors.` And I don't quite understand why, does default not capture the rest of the options?

Comment (Chemaclass) [reply src/Router/Entities/RouteParams.php:60]: Remove lines 8,9 from phpstan.neon

File: tests/Feature/Router/Fixtures/FakeController.php [hunk 3]
```diff
@@ -40,4 +40,17 @@ public function manyParamsAction(string $firstParam, string $secondParam, string
...
+     * @param mixed $param
+     */
+    public function nonTypedParam($param): string
+    {
+        return 'I AM ERROR!';
+    }
+
+    public function nonScalarParam(array $param): string
+    {
+        return 'I AM ERROR!';
```
Comment (Chemaclass) [inline tests/Feature/Router/Fixtures/FakeController.php:54]: I am *an* error

File: tests/Feature/Router/Fixtures/FakeController.php [hunk 4]
```diff
@@ -40,4 +40,17 @@ public function manyParamsAction(string $firstParam, string $secondParam, string
     {
         return "The params are '{$firstParam}', '{$secondParam}' and '{$thirdParam}'!";
     }
+
+    /**
+     * @param mixed $param
+     */
+    public function nonTypedParam($param): string
+    {
+        return 'I AM ERROR!';
```
Comment (Chemaclass) [inline tests/Feature/Router/Fixtures/FakeController.php:49]: I am **an** error

**Review by Chemaclass — APPROVED** [review]

File: infection.json5
Comment (Chemaclass) [inline infection.json5:1]: Is the extension `json5` correct?

File: tests/Feature/Router/Fixtures/FakeController.php (hunk 3 above)
Comment (antonio-gg-dev) [reply tests/Feature/Router/Fixtures/FakeController.php:54]: Nope, I AM ERROR
imagen

File: infection.json5
Comment (antonio-gg-dev) [reply infection.json5:1]: Yes, It's a json standard compatible with JavaScript syntax, I don't know why they have decided to use it for this library but I am not going to do the opposite.

File: tests/Feature/Router/Fixtures/FakeController.php (hunk 3 above)
Comment (Chemaclass) [reply tests/Feature/Router/Fixtures/FakeController.php:54]: Alles klar

**Comment by antonio-gg-dev** [comment]
You might want to take a look at this: https://infection.github.io/guide/mutation-badge.html

**Comment by Chemaclass** [comment]
I am using macos, and the `$(nproc)` that you are using in the composer script is unknown. 
Could we find another command? Or maybe using a raw-int, @Tito-Kati ?
Screenshot 2023-04-30 at 13 06 44

Using `--threads=4` takes: 17sec
Using `--threads=8` takes: 8sec

I would suggest to use 8 threads 😄

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -115,3 +115,32 @@ jobs:
...
+        with:
+          path: ${{ steps.composer-cache.outputs.dir }}
+          key: ${{ runner.os }}-composer-${{ hashFiles('**/composer.lock') }}
+          restore-keys: ${{ runner.os }}-composer-
+
+      - name: Install dependencies
+        run: composer install --no-interaction --no-ansi --no-progress
+
+      - name: Run mutation tests
+        run: ./vendor/bin/infection  --show-mutations --threads=$(nproc) --min-msi=100 --min-covered-msi=100
```
Comment (maks-rafalko) [inline .github/workflows/ci.yml:146]: You can use `--threads=max` that will automatically get the number of cores on any OS

Comment (antonio-gg-dev) [reply .github/workflows/ci.yml:146]: Thank you very much @maks-rafalko! The truth is that we discovered it yesterday while we were adding it to Gacela too.

Comment (Chemaclass) [reply .github/workflows/ci.yml:146]: Thanks, @maks-rafalko. I already changed it here https://github.com/gacela-project/router/commit/84df2cc5be4bb6e2ed8ebfa89e006a631cba931c
